### ISSUE TITLE
[Feature Request] Add NEAT (Nash-Equilibrium Adaptive Training) optimizer to keras.optimizers

### ISSUE THREAD

**Issue by ItCodinTime** [keras-team/keras#22639]
## Summary

This issue proposes adding **NEAT** (Nash-Equilibrium Adaptive Training) as a first-class optimizer in `keras.optimizers`. NEAT is a conflict-aware momentum optimizer that introduces a lightweight Nash-inspired correction term to suppress gradient components that directly conflict with the optimizer's own momentum history. The result is a smoother, more stable update rule that degrades gracefully to standard SGD-with-momentum when no conflict is detected.

I am the author of the reference implementation at https://github.com/ItCodinTime/NEAT and am prepared to deliver a complete, review-ready PR against `keras-team/keras:master` following all contribution guidelines.

---

## Motivation

Every optimizer in `keras.optimizers` today — SGD, Adam, AdamW, Adagrad, RMSprop, Nadam, Ftrl, Adamax — shares the same fundamental assumption: the gradient at step `t` is always the correct direction to move. None of them model **directional conflict** between the current gradient and the optimizer's own accumulated signal.

In practice, conflict is common:
- **Noisy or oscillatory loss surfaces** (e.g. small-batch training, transformers with warm-up)
- **Multi-task / multi-domain fine-tuning** where different data sources pull in opposing directions
- **Recurrent and sequence models** where gradients change direction across time steps
- **Adversarial or contrastive objectives** where the loss landscape is intentionally non-convex

When conflict occurs, a standard optimizer blindly accumulates the conflicting gradient into momentum, causing oscillation and slower convergence. NEAT detects conflict via the cosine similarity between `g_t` and the opponent proxy `o_t`, and removes the conflicting projection before accumulating momentum.

This is directly analogous to how **PCGrad** (Yu et al., NeurIPS 2020) and **CAGrad** (Liu et al., NeurIPS 2021) handle multi-task gradient conflict — but NEAT does it **within a single optimizer pass** with no per-task gradient storage, making it usable via `model.compile()` with zero changes to the training loop.

---

## Algorithm

For parameter tensor `θ`, gradient `g_t`, opponent proxy `o_t` (default: previous momentum `m_{t-1}`), correction strength `α`, and momentum decay `β`:

```
conflict_t  = relu( -cos(g_t, o_t) )           # ∈ [0, 1]; 0 when aligned
proj_t      = <g_t, o_t> / (‖o_t‖² + ε) · o_t  # projection of g onto o
nce_t       = -α · conflict_t · proj_t          # Nash Conflict-correction
u_t         = g_t + nce_t                        # corrected gradient
m_t         = β · m_{t-1} + (1-β) · u_t         # momentum accumulation
θ_{t+1}     = θ_t - lr · m_t                    # parameter update
```

**Key properties:**
- When `conflict_t = 0` (gradient and opponent are aligned or orthogonal), `nce_t = 0` and the update is **identical to SGD with momentum**. No regression on tasks where conflict is absent.
- The correction magnitude is bounded: `‖nce_t‖ ≤ nce_clip_ratio · ‖g_t‖` by construction, so it cannot destabilize training.
- The opponent proxy is configurable: `momentum` (default), `previous_gradient`, or `gradient_ema`.
- A warmup gate (`correction_warmup_steps`) and a conflict threshold gate (`conflict_threshold`) allow safe deployment when early gradients are noisy.

---

## Proposed API

```python
keras.optimizers.Neat(
    learning_rate=0.001,
    alpha=0.25,                    # conflict correction strength ∈ [0, 1]
    beta=0.9,                      # momentum decay
    nce_mode="projection",         # "projection" | "cosine"
    opponent_source="momentum",    # "momentum" | "previous_gradient" | "gradient_ema"
    opponent_ema_decay=0.9,        # used when opponent_source="gradient_ema"
    nce_clip_ratio=1.0,            # max ‖nce‖ / ‖g‖
    correction_warmup_steps=0,     # suppress correction for first N steps
    conflict_threshold=0.0,        # only apply correction when conflict > threshold
    weight_decay=None,             # decoupled L2 weight decay (AdamW-style)
    eps=1e-8,
)
```

**Usage (identical to any other Keras optimizer):**
```python
import keras

model = keras.Sequential([
    keras.layers.Dense(128, activation="relu"),
    keras.layers.Dense(10),
])

model.compile(
    optimizer=keras.optimizers.Neat(learning_rate=1e-3, alpha=0.25),
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
    metrics=["accuracy"],
)

model.fit(x_train, y_train, epochs=10)
```

The optimizer is fully serializable via `get_config()` / `from_config()` and compatible with all three Keras backends (TensorFlow, JAX, PyTorch).

---

## Relationship to Existing Optimizers

| Property | SGD | Adam | **NEAT** |
|---|---|---|---|
| Momentum | Yes | EMA (1st moment) | Yes |
| Adaptive LR per param | No | Yes (2nd moment) | No (base version) |
| Conflict detection | No | No | **Yes** |
| Decoupled weight decay | No | AdamW variant | Yes (built-in) |
| Reduces to SGD when no conflict | — | No | **Yes** |
| Extra memory per param | 0 | 1 tensor | 1–3 tensors (m, prev_g, g_ema) |

NEAT occupies a distinct position: it is not a second-moment method and does not adapt the learning rate per parameter. Instead it modifies **which direction** the gradient pushes, making it orthogonal to and composable with adaptive methods.

---

## Benchmarks

The reference repository includes a reproducible Keras MLP benchmark on the `sklearn` digits dataset (8x8 images, 10 classes, 3 seeds each):

| Optimizer | Mean Test Accuracy (20 epochs) |
|---|---|
| SGD (momentum=0.9) | 97.04% |
| Adam | 96.85% |
| AdamW | 96.85% |
| **NEAT (alpha=0.25, projection)** | **94.72%** |

NEAT trails Adam/SGD on this smooth, low-noise benchmark. This is **expected and honest**: the conflict correction has a mean ratio of only `0.00385` on this task (the gradients are largely aligned), so NEAT is effectively running as SGD with a small overhead. The optimizer's advantage surfaces on tasks with genuine gradient conflict — multi-task objectives, noisy fine-tuning, and oscillatory landscapes — which are precisely the use cases not well-served by the current optimizer suite.

Diagnostic tooling (`diagnostic_snapshot()`) is included in the reference implementation to measure the live conflict ratio during training, allowing users to verify whether their task benefits from the correction.

---

## Implementation Plan

If this proposal is accepted, I will deliver:

1. **`keras/src/optimizers/neat.py`** — Full `Neat(optimizer.Optimizer)` subclass following the exact same structure as `adam.py`, `adamw.py`, and `sgd.py` with:
   - `build()` creating `momentum`, `prev_grad`, and `grad_ema` slot variables
   - `update_step(gradient, variable, learning_rate)` implementing the update rule above, using only `keras.src.ops` (backend-agnostic)
   - `get_config()` / `from_config()` for full serialization
   - A complete NumPy-style docstring with Args, math description, and usage examples
   - Input validation with clear error messages

2. **`keras/src/optimizers/__init__.py`** — Add `Neat` to the exports

3. **`keras/api/_v2/keras/optimizers/__init__.py`** — Public API export

4. **`tests/optimizers/neat_test.py`** — Unit tests covering:
   - `alpha=0.0` reduces to SGD with momentum (numerical equality)
   - Correction is bounded by `nce_clip_ratio * ‖g‖`
   - Warmup gate: zero correction for first `N` steps
   - Conflict threshold gate: correction suppressed below threshold
   - `get_config()` round-trip serialization
   - `build()` slot variable shapes match variable shapes
   - Convergence on quadratic loss (x² → 0) within 100 steps
   - All three `opponent_source` modes run without error
   - Weight decay applies parameter shrinkage correctly

5. **Documentation** — Docstring follows the same format as `Adam` and will be auto-included in the API reference.

All code will use only `keras.src.ops` (no backend-specific calls), will pass `ruff check` and `ruff format`, and will be tested against the full Keras CI matrix.

---

## Reference

- Reference implementation: https://github.com/ItCodinTime/NEAT (Apache-2.0)
- Math specification: https://github.com/ItCodinTime/NEAT/blob/main/docs/research/math-spec.md
- Related work: PCGrad (Yu et al., NeurIPS 2020), CAGrad (Liu et al., NeurIPS 2021), Conflict-Averse Gradient Descent

---

I would appreciate any feedback on the proposed API surface before submitting the PR. Happy to adjust the parameter names, defaults, or scope based on maintainer guidance.

**Comment by hertschuh** [comment]
@ItCodinTime 

Thanks for writing this proposal. Was this published in some research paper?

**Comment by ItCodinTime** [comment]
Thank you for the response @hertschuh !

It was published to the Texas Junior Academy of Science and Texas State Science and Engineering Fair, was done in association with the University of North Texas!

Would love to make this avaialble to developers around the world!

[TJAS (3).pdf](https://github.com/user-attachments/files/26641301/TJAS.3.pdf)

**Comment by MarcosAsh** [comment]
Hi this looks quite interesting but I have a few questions about the proposal: 

1. How does this compare to existing gradient conflict mitigation methods like PCGrad? The core mechanism seems similar.

2. The benchmarks show NEAT underperforming SGD/Adam on the sklearn digits task. Can you share results on standard benchmarks like ImageNet or GLUE where the gradient conflicts you describe actually occur?

3. The paper mentions 28% faster convergence on a 1B parameter model, but the hardware listed (T4 16GB) seems insufficient for that scale. Can you clarify the experimental setup?

4. Also isnt the novel approach just SGD with momentum with a conflict correction term?

**Comment by ItCodinTime** [comment]
Hi! I don't know how but this notif got lost in my inbox, sorry to keep yall waiting

1. PCGrad and related techniques are designed for explicit multi task setups ie they compute per task gradients and resolve conflicts by projecting one tasks gradient onto the normal plane of another, which effectively changes the optimization objective when gradients disagree. NEAT instead assumes a single combined loss  and treats conflicts as a form of stochastic noise in the gradient field.

2. I'll work on this because from what ive seen its a pretty bulky test, and I'll need to contact the university again to reuse their hardware. But lightweight checks and their results are built into the readme and in their own folder

3. The 28% faster convergence claim comes from a controlled study on a 1b param model testing in a control setting with teh specified gpu, the T4 16GB line refers to the per GPU device, not training a full 1B parameter model e2e on a single card. If the wording was misleading I apologize and will fix that in another release of the paper

4. The novel part is the conflict aware signal NEAT gives, instead of treating all gradients equally, NEAT explicitly tracks when successive steps ) disagree and uses that information to adapt learning rates / momentum only in those high‑conflict regions. Standard momentum has no such notion of “conflicting directions" smoothening gradients over time.

Sorry for the delay!

**Comment by ItCodinTime** [comment]
Any updates guys? @MarcosAsh @sachinprasadhs @dhantule  would love to have this accessible the public

**Comment by hertschuh** [comment]
Thanks for reaching out and providing the details.

As a general rule, we add to Keras techniques that are not only state of the art but also widely accepted by the ML community. It does not appear that NEAT has wide adoption right now, but please check back once it does.

That being said, if you have a working Keras version in your own package, we can link it here: https://keras.io/getting_started/ecosystem/

**Comment by ItCodinTime** [comment]
Thank you for the reply @hertschuh!

Would love to have this in the ecosystem, can I get information for the packaging and documentation requirements and what other reference docs I would need to submit to get on the ecosystem? 


TYSM!

**Comment by ItCodinTime** [comment]
@hertschuh @MarcosAsh @sachinprasadhs @dhantule 

Sorry for the ping, but could I get an update on the requirements to get NEAT into the Keras Ecosystem?

Thank you for the opportunity!
Goutham KR

**Comment by hertschuh** [comment]
You can send a PR that adds to this file:
https://github.com/keras-team/keras-io/blob/master/templates/getting_started/ecosystem.md?plain=1

It will show up here: https://keras.io/getting_started/ecosystem/

**Comment by ItCodinTime** [comment]
Thank you!

**Comment by ItCodinTime** [comment]
Just PRd!
